File: supabase/functions/_shared/keyframePlan.ts
// Lane A — keyframe planning (LOCKED video garment-swap architecture, §3).
//
// Pure helpers, no Deno / DOM globals, so vitest exercises them like the other
// _shared pure modules (frameSwap.ts, videoCompose.ts).
//
// The locked lane approves a few product-accurate Grok HERO keyframes — roughly
// every 12–24 frames, PLUS one at every pose change / scene cut — then
// PROPAGATES the approved garment across the intermediate frames. This module
// answers two questions the propagation step needs:
//
//   1. planKeyframes(...)  — WHICH frames become Grok keyframes.
//   2. assignSegments(...) — for every frame, which two keyframes bracket it
//                            (the propagation source(s)) and where it sits
//                            between them (blend weight t).
//
// See docs/VIDEO_SWAP_ARCHITECTURE.md §3. Keyframe generation itself reuses the
// proven grok-image-garment-proxy — this module only decides the plan.

/** Locked cadence band: a scheduled keyframe every 12–24 frames (§3). */
export const MIN_KEYFRAME_CADENCE = 12;
export const MAX_KEYFRAME_CADENCE = 24;
export const DEFAULT_KEYFRAME_CADENCE = 18;

export interface KeyframePlanOptions {
  /** Total extracted frames (indices 0..frameCount-1). */
  frameCount: number;
  /**
   * Scheduled cadence in frames. Clamped to [12, 24] (the locked band). A
   * keyframe is scheduled at 0, cadence, 2·cadence, … plus the final frame.
   */
  cadenceFrames?: number;
  /**
   * Forced anchor indices — pose changes / scene cuts detected upstream (motion
   * analysis, shot-boundary detection, or re-anchor from a flow break). These
   * ALWAYS become keyframes regardless of cadence (§3: "and at every pose
   * change / scene cut").
   */
  forcedAnchors?: number[];
}

/** Clamp a cadence into the locked 12–24 band. Non-finite → the default. */
export function clampCadence(cadence: number | undefined): number {
  if (!Number.isFinite(cadence)) return DEFAULT_KEYFRAME_CADENCE;
  return Math.min(
    MAX_KEYFRAME_CADENCE,
    Math.max(MIN_KEYFRAME_CADENCE, Math.round(cadence as number)),
  );
}

/**
 * Decide which frame indices become Grok hero keyframes.
 *
 * Always includes the first frame (0) and the last (frameCount-1) so every
 * intermediate frame is bracketed by two real keyframes — propagation is never
 * asked to extrapolate past the ends. Scheduled cadence keyframes and forced
 * anchors are merged, de-duplicated, and returned sorted ascending.
 */
export function planKeyframes(opts: KeyframePlanOptions): number[] {
  const frameCount = Math.floor(opts.frameCount);
  if (!(frameCount > 0)) throw new Error(`planKeyframes: invalid frameCount ${opts.frameCount}`);
  if (frameCount === 1) return [0];

  const cadence = clampCadence(opts.cadenceFrames);
  const set = new Set<number>();
  set.add(0);
  set.add(frameCount - 1);
  for (let i = cadence; i < frameCount - 1; i += cadence) set.add(i);

  for (const raw of opts.forcedAnchors ?? []) {
    const a = Math.floor(raw);
    if (Number.isFinite(a) && a >= 0 && a < frameCount) set.add(a);
  }

  return [...set].sort((a, b) => a - b);
}

export interface FrameSegment {
  /** The frame this record describes. */
  index: number;
  /** True when this frame is itself an approved keyframe (use it directly). */
  isKeyframe: boolean;
  /** Nearest keyframe at or before `index`. */
  prevKeyframe: number;
  /** Nearest keyframe at or after `index`. */
  nextKeyframe: number;
  /**
   * Normalised position of `index` within [prevKeyframe, nextKeyframe], 0..1.
   * 0 at prevKeyframe, 1 at nextKeyframe. A keyframe reports 0. Used as the
   * blend weight when a propagation engine warps from both bracketing
   * keyframes toward the middle.
   */
  t: number;
  /** Which bracketing keyframe is closer — the primary propagation source. */
  nearestKeyframe: number;
}

/**
 * For every frame 0..frameCount-1, resolve the two keyframes that bracket it and
 * where it sits between them. `keyframeIndices` must be sorted ascending and
 * cover both ends (as planKeyframes guarantees).
 */
export function assignSegments(keyframeIndices: number[], frameCount: number): FrameSegment[] {
  const n = Math.floor(frameCount);
  if (!(n > 0)) throw new Error(`assignSegments: invalid frameCount ${frameCount}`);
  const kfs = [...keyframeIndices]
    .filter((k) => Number.isFinite(k) && k >= 0 && k < n)
    .sort((a, b) => a - b);
  if (kfs.length === 0) throw new Error("assignSegments: no keyframes");
  if (kfs[0] !== 0) throw new Error("assignSegments: keyframes must include frame 0");
  if (kfs[kfs.length - 1] !== n - 1)
    throw new Error("assignSegments: keyframes must include the final frame");

  const kfSet = new Set(kfs);
  const out: FrameSegment[] = [];
  let seg = 0; // index into kfs of the current lower bracket

  for (let i = 0; i < n; i++) {
    // Advance so kfs[seg] <= i < kfs[seg+1] (or i == last keyframe).
    while (seg + 1 < kfs.length && kfs[seg + 1] <= i) seg++;
    const prev = kfs[seg];
    const next = seg + 1 < kfs.length ? kfs[seg + 1] : prev;
    const span = next - prev;
    const t = span > 0 ? (i - prev) / span : 0;
    const isKeyframe = kfSet.has(i);
    const nearest = t <= 0.5 ? prev : next;
    out.push({
      index: i,
      isKeyframe,
      prevKeyframe: prev,
      nextKeyframe: next,
      t,
      nearestKeyframe: nearest,
    });
  }
  return out;
}

/** Intermediate (non-keyframe) frames only — the set the propagation step fills. */
export function intermediateFrames(segments: FrameSegment[]): FrameSegment[] {
  return segments.filter((s) => !s.isKeyframe);
}

File: workers/warp_worker/keyframe_plan.py
"""Keyframe / anchor planning — Python port of ``_shared/keyframePlan.ts``.

Decides WHICH frames are anchors and, for every frame, which two anchors bracket
it (the bidirectional propagation sources) and where it sits between them
(``t``). Ends are always anchored so nothing extrapolates past a real anchor.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable, List, Sequence

from .config import DEFAULT_KEYFRAME_CADENCE, MAX_KEYFRAME_CADENCE, MIN_KEYFRAME_CADENCE


def clamp_cadence(cadence) -> int:
    try:
        c = float(cadence)
    except (TypeError, ValueError):
        return DEFAULT_KEYFRAME_CADENCE
    if c != c:  # NaN
        return DEFAULT_KEYFRAME_CADENCE
    return max(MIN_KEYFRAME_CADENCE, min(MAX_KEYFRAME_CADENCE, round(c)))


def plan_keyframes(
    frame_count: int,
    cadence_frames: int | None = None,
    forced_anchors: Iterable[int] = (),
) -> List[int]:
    """Anchor indices: always 0 and frame_count-1, scheduled cadence anchors, plus
    forced anchors (pose changes / re-anchors). Sorted, de-duplicated."""
    n = int(frame_count)
    if n <= 0:
        raise ValueError(f"plan_keyframes: invalid frame_count {frame_count}")
    if n == 1:
        return [0]
    cadence = clamp_cadence(cadence_frames)
    s = {0, n - 1}
    i = cadence
    while i < n - 1:
        s.add(i)
        i += cadence
    for raw in forced_anchors:
        a = int(raw)
        if 0 <= a < n:
            s.add(a)
    return sorted(s)


@dataclass
class FrameSegment:
    index: int
    is_keyframe: bool
    prev_keyframe: int
    next_keyframe: int
    t: float
    nearest_keyframe: int


def assign_segments(keyframe_indices: Sequence[int], frame_count: int) -> List[FrameSegment]:
    """For every frame resolve its bracketing anchors + blend weight ``t``.
    ``keyframe_indices`` must be sorted and cover both ends."""
    n = int(frame_count)
    if n <= 0:
        raise ValueError(f"assign_segments: invalid frame_count {frame_count}")
    kfs = sorted(k for k in keyframe_indices if 0 <= k < n)
    if not kfs:
        raise ValueError("assign_segments: no keyframes")
    if kfs[0] != 0:
        raise ValueError("assign_segments: keyframes must include frame 0")
    if kfs[-1] != n - 1:
        raise ValueError("assign_segments: keyframes must include the final frame")
    kf_set = set(kfs)
    out: List[FrameSegment] = []
    seg = 0
    for i in range(n):
        while seg + 1 < len(kfs) and kfs[seg + 1] <= i:
            seg += 1
        prev = kfs[seg]
        nxt = kfs[seg + 1] if seg + 1 < len(kfs) else prev
        span = nxt - prev
        t = (i - prev) / span if span > 0 else 0.0
        nearest = prev if t <= 0.5 else nxt
        out.append(
            FrameSegment(
                index=i,
                is_keyframe=i in kf_set,
                prev_keyframe=prev,
                next_keyframe=nxt,
                t=t,
                nearest_keyframe=nearest,
            )
        )
    return out


def intermediate_frames(segments: Sequence[FrameSegment]) -> List[FrameSegment]:
    return [s for s in segments if not s.is_keyframe]
